quil-rs/src/expression.rs

```diff
@@ -253,6 +253,537 @@ fn is_small(x: f64) -> bool {
     x.abs() < 1e-16
 }
 
+/// A submodule whose sole purpose is to contain complex machinery for simplifying [`Expression`]s.
+mod simplification {
+    use super::{
+        format_complex, is_small, Expression, ExpressionFunction, FunctionCallExpression,
+        InfixExpression, InfixOperator, MemoryReference, PrefixExpression, PrefixOperator,
+    };
+    use egg::{define_language, rewrite as rw, FromOp, Id, Language, RecExpr};
+    use ordered_float::OrderedFloat;
+    use std::{
+        ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+        str::FromStr,
+    };
+    use symbolic_expressions::{parser, IntoSexp, Sexp};
+
+    /// Simplify an [`Expression`]:
+    /// - turn it into a [`Sexp`],
+    /// - parse that into a [`RecExpr<Expr>`],
+    /// - let [`egg`] simplify the recursive expression as best as it can,
+    /// - and turn that back into an [`Expression`] (by way of another [`Sexp`]).
+    pub(super) fn run(expression: &Expression) -> Result<Expression, SimplificationError> {
+        let sexp = expression.into_sexp();
+        let recexpr = parse_expr(&sexp)?;
+        let runner = egg::Runner::default()
+            .with_expr(&recexpr)
+            .run(&make_rules());
+        let root = runner.roots[0];
+        let (_, best) = egg::Extractor::new(&runner.egraph, egg::AstSize).find_best(root);
+        let simpler_sexp = parser::parse_str(&best.to_string())
+            .map_err(|e| SimplificationError::SexpFromRecExpr(format!("{e:#?}")))?;
```

Comment:
Is there a specific reason for stringifying the error instead of including the error object?

Reply:
I wasn't sure about bubbling up a `SexpError` from `symbolic_expressions` by way of `egg`. If we don't mind that popping up, I'll change this to use `#[from] SexpError`.